include/rcpputils/filesystem_helper.hpp

```diff
@@ -352,6 +360,40 @@ RCPPUTILS_PUBLIC bool operator!=(const path & a, const path & b);
 */
 RCPPUTILS_PUBLIC std::ostream & operator<<(std::ostream & os, const path & p);
 
+// remove warning suppression
+#if !defined(_WIN32)
+# pragma GCC diagnostic pop
+#else  // !defined(_WIN32)
+# pragma warning(pop)
+#endif
+
+/**
+ * \brief Get a path to a location in the temporary directory, if it's available.
+ *
+ * This does not create any directories.
+ * On Windows, this uses "GetTempPathA"
+ * On non-Windows, this prefers the environment variable TMPDIR, falling back to /tmp
+ *
+ * \return A path to a directory for storing temporary files and directories.
+ */
+RCPPUTILS_PUBLIC std::filesystem::path temporary_directory_path();
+
+/**
+ * \brief Construct a uniquely named temporary directory, in "parent", with format base_nameXXXXXX
+ *
+ * The output, if successful, is guaranteed to be a newly-created directory.
+ * The underlying implementation keeps generating paths until one that does not exist is found.
+ * This guarantees that there will be no existing files in the returned directory.
+ *
+ * \param[in] base_name User-specified portion of the created directory
+ * \param[in] parent_path The parent path of the directory that will be created
+ * \return A path to a newly-created directory with base_name and a 6-character unique suffix
+ *
+ * \throws std::system_error If any OS APIs do not succeed.
+ */
+RCPPUTILS_PUBLIC std::filesystem::path create_temporary_directory(
+  const std::string & base_name,
+  const std::filesystem::path & parent_path = temporary_directory_path());
```

Comment:
I poked at this a while today with @mjcarroll , and tried to come up with a nicer way to do this.  We really couldn't figure out a better way than making a new function name.

However, I pulled down all of the source code released into Rolling.  And I found that `rcpputils::fs::temp_directory_path` and `rcpputils::fs::create_temp_directory` aren't used anywhere outside of the core.  In the core, they are only used in:

* `rcpputils` - the implementation and the tests
* `rosbag2_compression` - some tests
* `rosbag2_cpp` - some tests
* `rosbag2_storage_mcap` - a test
* `rosbag2_storage_sqlite3` - a test
* `rosbag2_transport` - a test
* `rosbag2_test_common` - a test
* `rcl_logging_spdlog` - a test

Thus, at the cost of only 3 PRs (this one, one in `rcl_logging`, and one in `rosbag2`), we can completely break this API.  I think we should do that.  In particular, I think we should remove `temporary_directory_path` and `create_temporary_directory`, and also remove `temp_directory_path`.  Then we change `create_temp_directory` to take in a `std::filesystem::path`, and also return a `std::filesystem::path`.  And then we update the tests here to use `std::filesystem::path`, as well as the tests in `rcl_logging` and `rosbag2`.  At the end of those 3 PRs (which will have to be merged at the same time), we'll have the API we want going forward.

Does that make sense to you @ahcorde?

Reply:
Actually, in point of fact, that is a bit enough change that I think we should do that on its own, separate from this PR.  Once that is in, we can revisit deprecating `rcpputils::fs::path` separately (i.e. this PR).

Reply:
Related PR https://github.com/ros2/rcpputils/pull/197